crates/bevy_pbr/src/light.rs

```diff
@@ -480,14 +481,61 @@ fn cluster_to_index(cluster_dimensions: UVec3, cluster: UVec3) -> usize {
     ((cluster.y * cluster_dimensions.x + cluster.x) * cluster_dimensions.z + cluster.z) as usize
 }
 
+// Sort point lights with shadows enabled first, then by a stable key so that the index
+// can be used to render at most `MAX_POINT_LIGHT_SHADOW_MAPS` point light shadows, and
+// we keep a stable set of lights visible
+pub(crate) fn point_light_order(
+    (entity_1, shadows_enabled_1): (&Entity, &bool),
+    (entity_2, shadows_enabled_2): (&Entity, &bool),
+) -> std::cmp::Ordering {
+    shadows_enabled_1
+        .cmp(shadows_enabled_2)
+        .reverse()
+        .then_with(|| entity_1.cmp(entity_2))
+}
+
 // NOTE: Run this before update_point_light_frusta!
 pub fn assign_lights_to_clusters(
     mut commands: Commands,
     mut global_lights: ResMut<VisiblePointLights>,
     mut views: Query<(Entity, &GlobalTransform, &Camera, &Frustum, &mut Clusters)>,
     lights: Query<(Entity, &GlobalTransform, &PointLight)>,
 ) {
-    let light_count = lights.iter().count();
+    let mut lights = lights.iter().collect::<Vec<_>>();
+
+    if lights.len() > MAX_POINT_LIGHTS {
+        lights.sort_by(|(entity_1, _, light_1), (entity_2, _, light_2)| {
+            point_light_order(
+                (entity_1, &light_1.shadows_enabled),
+                (entity_2, &light_2.shadows_enabled),
+            )
+        });
+
+        // check each light against each view's frustum, keep only those that affect at least one of our views
+        let frusta: Vec<_> = views.iter().map(|(_, _, _, frustum, _)| *frustum).collect();
+        lights = lights
+            .into_iter()
+            .filter(|(_, light_transform, light)| {
+                let light_sphere = Sphere {
+                    center: light_transform.translation,
+                    radius: light.range,
+                };
+
+                frusta
+                    .iter()
+                    .any(|frustum| frustum.intersects_sphere(&light_sphere))
+            })
+            // take one extra light to check if we should emit the warning
+            .take(MAX_POINT_LIGHTS + 1)
+            .collect();
+
+        if lights.len() > MAX_POINT_LIGHTS {
+            warn!("MAX_POINT_LIGHTS exceeded");
```

Comment:
It would be good to note what the MAX_POINT_LIGHTS limit is to make this more actionable.

Reply:
done